public/app/features/alerting/unified/components/notification-policies/NotificationPoliciesList.tsx

```diff
@@ -112,56 +126,38 @@ export const NotificationPoliciesList = () => {
 
   const refetchPolicies = () => {
     refetchNotificationPolicyRoute();
-    updateRouteTree.reset();
+    updateExistingNotificationPolicy.reset();
+    deleteNotificationPolicy.reset();
+    addNotificationPolicy.reset();
   };
 
-  function handleSave(partialRoute: Partial<FormAmRoute>) {
-    if (!rootRoute) {
-      return;
-    }
-
-    const newRouteTree = mergePartialAmRouteWithRouteTree(selectedAlertmanager ?? '', partialRoute, rootRoute);
-    updateRouteTree.execute(newRouteTree);
+  async function handleUpdate(partialRoute: Partial<FormAmRoute>) {
+    await updateExistingNotificationPolicy.execute(partialRoute);
+    handleActionResult({ error: updateExistingNotificationPolicyState.error });
   }
 
-  function handleDelete(route: RouteWithID) {
-    if (!rootRoute) {
-      return;
-    }
-    const newRouteTree = omitRouteFromRouteTree(route, rootRoute);
-    updateRouteTree.execute(newRouteTree);
+  async function handleDelete(route: RouteWithID) {
+    await deleteNotificationPolicy.execute(route.id);
+    handleActionResult({ error: deleteNotificationPolicyState.error });
   }
 
-  function handleAdd(partialRoute: Partial<FormAmRoute>, referenceRoute: RouteWithID, insertPosition: InsertPosition) {
-    if (!rootRoute) {
-      return;
-    }
-
-    const newRouteTree = addRouteToReferenceRoute(
-      selectedAlertmanager ?? '',
+  async function handleAdd(
+    partialRoute: Partial<FormAmRoute>,
+    referenceRoute: RouteWithID,
+    insertPosition: InsertPosition
+  ) {
+    await addNotificationPolicy.execute({
       partialRoute,
-      referenceRoute,
-      rootRoute,
-      insertPosition
-    );
-    updateRouteTree.execute(newRouteTree);
+      referenceRouteIdentifier: referenceRoute.id,
+      insertPosition,
+    });
+    handleActionResult({ error: addNotificationPolicyState.error });
   }
 
-  // this function will make the HTTP request and tracks the state of the request
-  const [updateRouteTree, updateRouteTreeState] = useAsync(async (routeTree: Route | RouteWithID) => {
-    if (!rootRoute) {
-      return;
+  function handleActionResult({ error }: { error?: Error }) {
+    if (!error) {
+      appNotification.success('Updated notification policies');
     }
-
-    // make sure we omit all IDs from our routes
-    const newRouteTree = cleanRouteIDs(routeTree);
-
-    const newTree = await updateNotificationPolicyRoute({
-      newRoute: newRouteTree,
-      oldRoute: defaultPolicy,
-    });
-
-    appNotification.success('Updated notification policies');
     if (selectedAlertmanager) {
       refetchAlertGroups();
```

Comment:
(pre-existing code, but) Could we make the mutation invalidate the tags provided for alert groups? So it auto-refetches?

Reply:
_technically_ there's no relationship between the alert groups and the policies – I'd put this aside for now and we can always revisit later :)